config: window-inherit-working-directory options

jparise opened the pull request:
`window-inherit-working-directory` was previously a boolean value. This change makes it a more granular option, allowing you to enable this behavior for one or more surface "kinds": split, tab, window.

The full behavior is only implemented for the embedded runtime and macOS app. The other runtimes still need their own plumbing to track surface kind creation (stubbed as `.window` by default in this change).

Resolves #1392

jparise committed:
config: window-inherit-working-directory options

`window-inherit-working-directory` was previously a boolean value. This
change makes it a more granular option, allowing you to enable this
behavior for one or more surface "kinds": split, tab, window.

The configuration value can be `true` (all), `false` (none), or a
comma-delimited string of kinds (e.g. `tab` or `split,window`).

The full behavior is only implemented for the embedded runtime and macOS
app. The other runtimes still need their own plumbing to track surface
kind creation (stubbed as `.window` by default in this change).

mitchellh reviewed (changes requested):
So far so good. Am I missing where the kind is actually set from the apprts?

jparise committed:
config: restore window-inherit-working-directory defaults

This setting previously defaulted to `true`, so flip all of our packed
booleans on by default.

Also, note that the granular values are only supported on macOS.

jparise committed:
config: use the standard 'packed struct' behavior

jparise committed:
config: switch to a Surface.Options-based approach

This introduces an ArenaAllocator that models the lifetime of the
Surface.Options structure from initial surface creation command to the
actual surface creation. This implementation is a little rough and
relies on an `anyopaque` pointer (because `extern struct`), but it
demonstrates a more apprt-centric approach.

Still only implemented for embedded apprt and macOS. The other apprts
will need their own plumbing to support this approach.

We can also chose to clean up newConfig() and call Config.shallowClone()
directly after this change.
